File: 프로그래머스/2/389479. 서버 증설 횟수/서버 증설 횟수.java
import java.util.*;

class Solution {
    public int solution(int[] players, int m, int k) {
        
        Deque<Integer> server = new ArrayDeque<>();
        int answer = 0;
        
        for(int i=0; i<players.length; i++){
            while(!server.isEmpty() && server.peek() < i){
                server.poll();
            }
            int need = players[i]/m;
            if(server.size() >= need) continue;
            int plus = need - server.size();
            for(int j=0; j<plus; j++){
                server.add(i+k-1);
                answer++;
            }
        }
    
        return answer;
    }
}

File: 프로그래머스/2/389479. 서버 증설 횟수/서버 증설 횟수.js
function solution(players, m, k) {
    
    const server = [];
    var answer = 0;
    
    for(let i=0; i<players.length; i++){
        while(server.length && server[0] < i){
            server.shift();
        }
        var need = Math.floor(players[i]/m);
        if(server.length >= need) continue;
        var plus = need - server.length;
        for(let j=0; j<plus; j++){
            server.push(i+k-1);
            answer++;
        }
    }
    
    return answer;
}